Known Issues and Workarounds page (#3193)
AxonFramework/AxonFramework: docs/old-reference-guide/modules/ROOT/pages/known-issues-and-workarounds.adoc

Especially given the fact Axon Framework will integrate with a wide variety of other tools, it becomes rough to deal with everything.

This page intends to provide the biggest "gotchas," with optional workarounds, which you can see below.

[NOTE]
.Where is issue XYZ?
====
This page can by no means contain **all** possible issues out there.
However, that does not mean we are not open for suggestions.

This documentation, just as the framework, is open source as well.
Hence, if you have found something problematic that you think should be added to the list, be sure to file an issue and (ideally!) a pull request.

You can make an enhancement request by clicking link:https://github.com/AxonFramework/AxonFramework/issues/new?assignees=&labels=Type%3A+Enhancement&projects=&template=2_enhancement_request.md[here].
When looking to provide a pull request, be sure to adjust link:https://github.com/AxonFramework/AxonFramework/blob/master/docs/old-reference-guide/modules/ROOT/pages/serialization.adochttps://github.com/AxonFramework/AxonFramework/blob/master/docs/old-reference-guide/modules/ROOT/pages/known-issues-and-workarounds.adoc[this file].
====

> == Relation database storage and serialized data columns

Reviewer: Typo: relation -> relational

When storing the serialized formats of, for example, xref:axon-framework-reference:events:event-processors/streaming.adoc#tracking-tokens[tokens], xref:axon-framework-reference:sagas:index.adoc[sagas], xref:axon-framework-reference:tuning:event-snapshots.adoc[snapshots], or xref:axon-framework-reference:events:infrastructure.adoc[events], it is good to know Axon uses the (B)LOB for these columns in your RDBMS of choice.

Some databases react unexpectedly to this, with a clear case being link:https://www.postgresql.org/[PostgreSQL].

> It uses a mechanism it calls link:https://wiki.postgresql.org/wiki/TOAST["TOAST"] to move data that's too large to another table.

Reviewer: I might be misreading the sources, but it looks like the blog's title is misleading.

TOAST is actually transparent for the user, and more powerful. This is automatically used for bytea (byte array) data types. For LOB datatypes, the OID approach is used, which is quite annoying, as it requires cleanup.
The Postgres driver for Hibernate maps `@Lob` annotated fields to LOB data types.

So we want more TOAST, not less.

Reviewer 2: It's both TOAST. However, one is eager and the other is lazy.

Hibernate maps to LOB columns automatically, which is eagerly TOASTing. This means that all token values are stored into the TOAST table of the token_entry table. 
Now, you can't really clean up the OID table (which we experienced first-hand in Inspector). In addition, it needs an extra read.
As tokens generally far from exceed 8KB (especially compressed), there's not reason to toast at all. Setting the datatype to bytea (lazy toasting) is thus basically leading to never toast :) 

For context, as I did some digging for inspector; To clean up OIDs, you have a special command. This makes a copy of all the OID values that still have a reference, reordens it, and then deleted all the values not in that table in the original table. It's horrible to manage :sweat:

Author: I have decided it would be better to simply add those changes to this PR. So, I did just that right now.

Hence, @abuijze and @CodeDrivenMitch, please review it again. Your opinion is much appreciated in this area.

For those cases, it would be wise to replace the TOAST behavior entirely.
The most straightforward way, is to enforce the use of the `bytea` column type; with some additional steps, of course.
The following link:https://www.axoniq.io/blog/axonframework-and-postgresql-without-toast[blog post] explains it in great detail, so be sure to check it out.

== Sequence generation issues with JPA Entities

For the JPA entities that require a generated sequence, Axon opted to use the default `@GeneratedValue` annotation back in 2016.
In the pre-Hibernate-6 age, this typically resulted in the construction of a single sequence generator that would be reused by several tables.
This can already be regarded as suboptimal, as frequently the sequence generator would be reused between tables.

Especially given the nature of the `global_index` column on the `domain_event_entry` table, which is used to keep the progress in xref:axon-framework-reference:events:event-processors/streaming.adoc[streaming event processors], a reused generator causes unexpected gaps in event processing.
Although Axon can deal with these gaps, they do make the queries to the `domain_event_entry` table rather bulky.

> Just as with any event storage solution, using a dedicate event store implementation like xref:axon-server-reference::index.adoc[Axon Server] will solve the problem entirely.

Reviewer 3: Did you mean to say `dedicated` instead of `dedicate`?

Author: You are entirely right! Thanks for spotting it. I'll adjust it accordingly.

= Known Issues and Workarounds
:navtitle: Known Issues and Workarounds

As with any tool, library, or framework, there are well-known issues you might find with Axon Framework.

> Especially since Axon Framework integrates with many other tools, it becomes rough to deal with everything.

Reviewer 4: ```suggestion
Since Axon Framework integrates with many other tools, it is challenging to find the perfect solution for every scenario.
```

This causes rows to contain an OID (object identifier) instead of the data directly.
Hibernate will default `@Lob` annotated fields to go through this TOAST feature, by mapping to the OID type.

Hence, all your tokens, sagas, snapshots, and events will automatically move to another table, with all added overhead coming with it.
Using xref:axon-server-reference::index.adoc[Axon Server] will, obviously, resolve this for your events and snapshots, as you would no longer use an RDBMS in that case.
However, even when using Axon Server, the predicament remains for tokens and sagas.

For those cases, it would be wise to adjust the Hibernate mapping.
The most straightforward way, is to enforce the use of the `bytea` column type; with some additional steps, of course.
For a full explanation and walkthrough, please check the xref:axon-framework-reference:tuning:rdbms-tuning.adoc#_postgresql_lob_annotated_columns_and_default_hibernate_mapping[PostegreSQL, LOB annotated columns, and default Hibernate mapping] tuning page.

== Sequence generation issues with JPA Entities

> For the JPA entities that require a generated sequence, Axon opted to use the default `@GeneratedValue` annotation back in 2016.

Reviewer 4: ```suggestion
For the JPA entities that require a generated sequence, we decided to use the default `@GeneratedValue` annotation back in 2016.
```

In the pre-Hibernate-6 age, this typically resulted in the construction of a single sequence generator that would be reused by several tables.

> This can already be regarded as suboptimal, as frequently the sequence generator would be reused between tables.

Reviewer 4: ```suggestion
This can be regarded as suboptimal, as frequently, the sequence generator would be reused between tables.
```

> Especially given the nature of the `global_index` column on the `domain_event_entry` table, which is used to keep the progress in xref:axon-framework-reference:events:event-processors/streaming.adoc[streaming event processors], a reused generator causes unexpected gaps in event processing.

Reviewer 4: ```suggestion
Especially given the nature of the `global_index` column on the `domain_event_entry` table, which is used to keep the progress in xref:axon-framework-reference:events:event-processors/streaming.adoc[streaming event processors], a reused generator, causes unexpected gaps in event processing.
```

> Although Axon can deal with these gaps, they do make the queries to the `domain_event_entry` table rather bulky.

Reviewer 4: ```suggestion
Although Axon Framework can deal with these gaps, they make the queries to the `domain_event_entry` table rather bulky.
```
